File: Behavioral/Lab22/Typescript/src/TemplateMethods/add_house_template.ts
import { Agency } from "../Entities/agency";
import { House } from "../Entities/house";
import { AbstractTemplate } from "./abstract_template";

export class AddHouseStrategy extends AbstractTemplate {
    constructor(...params: string[]) {
        super(...params);
    }

    protected override addEmpty(): string {
        const agency = Agency.GetInstance();
        agency.AddEstate(new House());
        return "Added empty new House";
    }

    protected override addEstate(): string {
        const agency = Agency.GetInstance();
        if (this.Params.length === 5) {
            agency.AddEstate(new House(
                this.Params[0],
                this.Params[1],
                +this.Params[2],
                +this.Params[3],
                this.Params[4]
            ));
            return "Added new House with args: " + this.Params.join(", ");
        }
        return this.returnError();
    }

    protected override returnError(): string {
        return "Failed to add new House with args: " + this.Params.join(",");
    }
}

File: Behavioral/Lab22/C#/TemplateMethods/AddHouseTemplate.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Entities;

namespace C_.TemplateMethods
{
    public class AddHouseTemplate : AbstractTemplate
    {
        public AddHouseTemplate(params string[] args) : base(args) {}

        protected override string AddEmpty()
        {
            var agency = Agency.GetInstance();
            agency.AddEstate(new House());
            return "Added empty new House";
        }

        protected override string AddEstate()
        {
            if(_params.Length != 5) 
                return ReturnError();
            var agency = Agency.GetInstance();
            agency.AddEstate(new House(_params[0], _params[1], Convert.ToDouble(_params[2]), Convert.ToDouble(_params[3]), _params[4]));
            return "Added new House with args: " + string.Join(", ", _params);
        }

        protected override string ReturnError()
        {
            return "Failed to add new House with args: " + string.Join(", ", _params);
        }
    }
}

